### PR TITLE
dns/eve: add 'HTTPS' type logging

### PR DESCRIPTION
**Pull request by hsadia538**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4751) ticket:
https://redmine.openinfosecfoundation.org/issues/4751

Describe changes:
- Add a new DNS record type to represent HTTPS

### CODE REVIEW

File: rust/src/dns/log.rs [hunk 1]
```diff
@@ -86,6 +86,7 @@ pub const LOG_URI        : u64 = BIT_U64!(59);
 
 pub const LOG_FORMAT_GROUPED  : u64 = BIT_U64!(60);
 pub const LOG_FORMAT_DETAILED : u64 = BIT_U64!(61);
+pub const LOG_HTTPS      : u64 = BIT_U64!(62);
```
Comment (inashivb) [inline rust/src/dns/log.rs:89]: Let's try to keep it the same number of spaces before `:` as above.

**Review by inashivb — CHANGES_REQUESTED** [review]
Hi @hsadia538 !
Good job understanding the issue. I feel that we are still missing a bunch of code to make it work.
Jotting down some hints, goals and todos for you.

**Goal:**
Make sure that in the dns logs, instead of `rrtype: 65`, we see `rrtype: HTTPS`.

**Hints:**
1. f you would check out any pre existing Suricata test (see any dns tests [here](https://github.com/OISF/suricata-verify/tree/master/tests)), you'll see that happen for a code defined RRType. e.g. `rrtype: A`. So, what you could do is follow all the places in the log.rs (and maybe others) where rrtype A is dealt with.
2. Your best friend: `git grep`
3. Bonus hint: Try to run grep on the entire codebase and not just the Rust part, you may find something interesting. 😉 

**Next todos:**
1. Create a [suricata-verify](https://github.com/oiSF/suricata-verify) test. You can use [createst.py](https://github.com/OISF/suricata-verify/blob/master/createst.py) for this.
Tip: Generally any issues that require changes in logging behavior should have a suricata-verify test. A good hint in this case is the presence of a pcap on the redmine ticket "proving" the issue.
2. Make sure Suricata logs `rrtype: HTTPS` like all the other defined rrtypes.

File: rust/src/dns/log.rs (hunk 1 above)
Comment (hsadia538) [reply rust/src/dns/log.rs:89]: Alright, I kept this based on the spacing on lines above line 87 due to its length. I will do this after completing the to do's

**Comment by hsadia538** [comment]
> Hi @hsadia538 ! Good job understanding the issue. I feel that we are still missing a bunch of code to make it work. Jotting down some hints, goals and todos for you.
> 
> **Goal:** Make sure that in the dns logs, instead of `rrtype: 65`, we see `rrtype: HTTPS`.
> 
> **Hints:**
> 
> 1. f you would check out any pre existing Suricata test (see any dns tests [here](https://github.com/OISF/suricata-verify/tree/master/tests)), you'll see that happen for a code defined RRType. e.g. `rrtype: A`. So, what you could do is follow all the places in the log.rs (and maybe others) where rrtype A is dealt with.
> 2. Your best friend: `git grep`
> 3. Bonus hint: Try to run grep on the entire codebase and not just the Rust part, you may find something interesting. 😉
> 
> **Next todos:**
> 
> 1. Create a [suricata-verify](https://github.com/oiSF/suricata-verify) test. You can use [createst.py](https://github.com/OISF/suricata-verify/blob/master/createst.py) for this.
>    Tip: Generally any issues that require changes in logging behavior should have a suricata-verify test. A good hint in this case is the presence of a pcap on the redmine ticket "proving" the issue.
> 2. Make sure Suricata logs `rrtype: HTTPS` like all the other defined rrtypes.

Thank you for the detailed review!
I think I will be coming back with more questions on this as I am trying to understand it.

File: rust/src/dns/log.rs (hunk 1 above)
Comment (inashivb) [reply rust/src/dns/log.rs:89]: I certainly did not check that. This seems to require some formatting. Let's leave it where it is as of now. Sorry about the trouble. 😅

**Comment by hsadia538** [comment]
> Hi @hsadia538 ! Good job understanding the issue. I feel that we are still missing a bunch of code to make it work. Jotting down some hints, goals and todos for you.
> 
> **Goal:** Make sure that in the dns logs, instead of `rrtype: 65`, we see `rrtype: HTTPS`.
> 
> **Hints:**
> 
> 1. f you would check out any pre existing Suricata test (see any dns tests [here](https://github.com/OISF/suricata-verify/tree/master/tests)), you'll see that happen for a code defined RRType. e.g. `rrtype: A`. So, what you could do is follow all the places in the log.rs (and maybe others) where rrtype A is dealt with.
> 2. Your best friend: `git grep`
> 3. Bonus hint: Try to run grep on the entire codebase and not just the Rust part, you may find something interesting. 😉
> 
> **Next todos:**
> 
> 1. Create a [suricata-verify](https://github.com/oiSF/suricata-verify) test. You can use [createst.py](https://github.com/OISF/suricata-verify/blob/master/createst.py) for this.
>    Tip: Generally any issues that require changes in logging behavior should have a suricata-verify test. A good hint in this case is the presence of a pcap on the redmine ticket "proving" the issue.
> 2. Make sure Suricata logs `rrtype: HTTPS` like all the other defined rrtypes.

I created an SV test based on a DNS test which is testing `rrtype:A`. I am confused about how it actually works when we run the test. Is it running Suricata on the `test input pcap` to validate test outputs? If yes then e.g if I comment out some code related to the above rrtype the test still passes. My expectation was that once I mess up the code the test will fail( I am doing `./configure`, `make` and `make install` after changing the code)

**Comment by jlucovsky** [comment]
> > Hi @hsadia538 ! Good job understanding the issue. I feel that we are still missing a bunch of code to make it work. Jotting down some hints, goals and todos for you.
> > **Goal:** Make sure that in the dns logs, instead of `rrtype: 65`, we see `rrtype: HTTPS`.
> > **Hints:**
> > 
> > 1. f you would check out any pre existing Suricata test (see any dns tests [here](https://github.com/OISF/suricata-verify/tree/master/tests)), you'll see that happen for a code defined RRType. e.g. `rrtype: A`. So, what you could do is follow all the places in the log.rs (and maybe others) where rrtype A is dealt with.
> > 2. Your best friend: `git grep`
> > 3. Bonus hint: Try to run grep on the entire codebase and not just the Rust part, you may find something interesting. 😉
> > 
> > **Next todos:**
> > 
> > 1. Create a [suricata-verify](https://github.com/oiSF/suricata-verify) test. You can use [createst.py](https://github.com/OISF/suricata-verify/blob/master/createst.py) for this.
> >    Tip: Generally any issues that require changes in logging behavior should have a suricata-verify test. A good hint in this case is the presence of a pcap on the redmine ticket "proving" the issue.
> > 2. Make sure Suricata logs `rrtype: HTTPS` like all the other defined rrtypes.
> 
> I created an SV test based on a DNS test which is testing `rrtype:A`. I am confused about how it actually works when we run the test. Is it running Suricata on the `test input pcap` to validate test outputs? If yes then e.g if I comment out some code related to the above rrtype the test still passes. My expectation was that once I mess up the code the test will fail( I am doing `./configure`, `make` and `make install` after changing the code)

`suricata-verify` will execute `suricata` (from the directory where you launched `suricata-verify`) for the test cases. You can do `suricata-verify --quiet my-test-case` to limit which tests cases it executes.

**Comment by inashivb** [comment]
> I created an SV test based on a DNS test which is testing `rrtype:A`. I am confused about how it actually works when we run the test. Is it running Suricata on the `test input pcap` to validate test outputs? 

Nicely answered by Jeff.

> If yes then e.g if I comment out some code related to the above rrtype the test still passes. My expectation was that once I mess up the code the test will fail( I am doing `./configure`, `make` and `make install` after changing the code)

Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.

Note: Just to clarify, I meant you should create a test for **this** issue with the pcap on the redmine ticket (4751). 😉 
And, the rrtype A example was to guide you which places in the code to look at to make changes for another rrtype (which is https in this case).
Please let us know if this is unclear.

**Comment by hsadia538** [comment]
> Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.
> 
> Note: Just to clarify, I meant you should create a test for **this** issue with the pcap on the redmine ticket (4751). 😉 And, the rrtype A example was to guide you which places in the code to look at to make changes for another rrtype (which is https in this case). Please let us know if this is unclear.

Yes, I was trying to execute tests for `rrtype: A` to track the whole process and understand how the tests are being executed. After this, I will move on to `rrtype:HTTPS`.

Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand.
image

**Comment by jufajardini** [comment]
> > Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.
> > Note: Just to clarify, I meant you should create a test for **this** issue with the pcap on the redmine ticket (4751). wink And, the rrtype A example was to guide you which places in the code to look at to make changes for another rrtype (which is https in this case). Please let us know if this is unclear.
> 
> Yes, I was trying to execute tests for `rrtype: A` to track the whole process and understand how the tests are being executed. After this, I will move on to `rrtype:HTTPS`.
> 
> Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand. image

Do you want to share a branch with the SV test you created, so we may have a look and possibly understand why is it still passing?

**Comment by inashivb** [comment]
> Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand. 

Make sure that the test you create has been created against `master` branch and then test it against your own branch. Otherwise, the test would also be incorrect. 😉

**Comment by hsadia538** [comment]
> > > Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.
> > > Note: Just to clarify, I meant you should create a test for **this** issue with the pcap on the redmine ticket (4751). wink And, the rrtype A example was to guide you which places in the code to look at to make changes for another rrtype (which is https in this case). Please let us know if this is unclear.
> > 
> > 
> > Yes, I was trying to execute tests for `rrtype: A` to track the whole process and understand how the tests are being executed. After this, I will move on to `rrtype:HTTPS`.
> > Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand. image
> 
> Do you want to share a branch with the SV test you created, so we may have a look and possibly understand why is it still passing?

Should I create a PR in s-v repo to get the review?

**Comment by hsadia538** [comment]
> > Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand.
> 
> Make sure that the test you create has been created against `master` branch and then test it against your own branch. Otherwise, the test would also be incorrect. 😉

I run the test on the master branch and then run it on my branch where I removed logic for `rrtye:A` but it is still passing on both. I will be sharing the sample test in a PR if that is okay.

**Comment by jufajardini** [comment]
> > > Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand.
> > 
> > 
> > Make sure that the test you create has been created against `master` branch and then test it against your own branch. Otherwise, the test would also be incorrect. wink
> 
> I run the test on the master branch and then run it on my branch where I removed logic for `rrtye:A` but it is still passing on both. I will be sharing the sample test in a PR if that is okay.

Yes, that's best as then we'll be able to visualize what's happening better. :)

**Comment by inashivb** [comment]
> Yes, I was trying to execute tests for `rrtype: A` to track the whole process and understand how the tests are being executed. After this, I will move on to `rrtype:HTTPS`.
> 
> Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand. image

I looked into this and this part of code is indeed not required for just the logging of an rrtype.

Notice the following from the previous comments. This part not causing your test to fail tells that the logging of `rrtype` is happening separately as and when a dns record comes in.

> > Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.

However, if you run all the dns tests with this code commented out, you'll see some of the tests fail which do depend on parsing of the rrtype A specifically.
Command:
```
python ../suricata-verify/run.py dns
```
You should see `dns-udp-eve-v2-txt`, `dns-udp-eve-v2-dig`, and some others fail.

**Comment by hsadia538** [comment]
> > Yes, I was trying to execute tests for `rrtype: A` to track the whole process and understand how the tests are being executed. After this, I will move on to `rrtype:HTTPS`.
> > Attaching the code for reference that I removed and reran the DNS tests but they were still passing and that's what I don't understand. image
> 
> I looked into this and this part of code is indeed not required for just the logging of an rrtype.
> 
> Notice the following from the previous comments. This part not causing your test to fail tells that the logging of `rrtype` is happening separately as and when a dns record comes in.
> 
> > > Well, it ideally should if the code you have commented out is actually being called while processing the pcap AND it causes a change in the logging.
> 
> However, if you run all the dns tests with this code commented out, you'll see some of the tests fail which do depend on parsing of the rrtype A specifically. Command:
> 
> ```
> python ../suricata-verify/run.py dns
> ```
> 
> You should see `dns-udp-eve-v2-txt`, `dns-udp-eve-v2-dig`, and some others fail.

Ok this makes more sense now. I think to completely understand it I will have to understand how the parsers are written and working actually. That is challenging and exciting as well. I have made some more changes to the code and need feedback to move forward so will. be opening another draft PR for that.

**Comment by hsadia538** [comment]
Followed by #8031

**Pull request closed without merging**